src/data_source.rs

```diff
@@ -0,0 +1,101 @@
+//! Holds implementation of odbc connection
+use super::{raw, Environment, Result, DiagRec, Error};
+use std;
+use std::marker::PhantomData;
+
+/// Represents a connection to an ODBC data source
+pub struct DataSource<'a> {
+    handle: raw::SQLHDBC,
+    // we use phantom data to tell the borrow checker that we need to keep the environment alive for
+    // the lifetime of the connection
+    env: PhantomData<&'a Environment>,
+}
+
+impl<'a> DataSource<'a> {
+    /// Connects to an ODBC data source
+    ///
+    /// # Arguments
+    /// * `env` - Environment used to allocate the data source handle.
+    /// * `dsn` - Data source name configured in the `odbc.ini` file!()
+    /// * `usr` - User identifier
+    /// * `pwd` - Authentication (usually password)
+    pub fn with_dsn_and_credentials<'b>(env: &'b mut Environment,
+                                        dsn: &str,
```

Comment:
May be it's better to use `Cow` instead of `&str` here (and in all public functions, actually) for better usability - see this [article](http://blog.jwilm.io/from-str-to-cow/) for example.